keyvi/include/keyvi/dictionary/fsa/internal/int_inner_weights_value_store.h

```diff
@@ -38,56 +38,63 @@ namespace dictionary {
 namespace fsa {
 namespace internal {
 
-/**
- * An IntValue store which uses the values as inner weights.
- */
-class IntInnerWeightsValueStore final : public IValueStoreWriter {
+class IntInnerWeightsValueStoreBase {
  public:
   typedef uint32_t value_t;
-  static const uint32_t no_value = 0;
   static const bool inner_weight = true;
 
-  using IValueStoreWriter::IValueStoreWriter;
+  static value_store_t GetValueStoreType() { return INT_WITH_WEIGHTS; }
 
-  // Because the copy ctor is "user declared", the default ctor is not
-  // generated by the compiler
-  IntInnerWeightsValueStore() = default;
-  IntInnerWeightsValueStore(const IntInnerWeightsValueStore& that) = delete;
-  IntInnerWeightsValueStore& operator=(IntInnerWeightsValueStore const&) = delete;
-
-  uint64_t GetValue(value_t value, bool* no_minimization) const {
+  uint64_t AddValue(value_t value, bool* no_minimization) const {
     TRACE("Value store value: %d", value);
 
     return value;
   }
-
   uint32_t GetWeightValue(value_t value) const { return value; }
 
-  uint64_t GetMergeValueId(size_t fileIndex, uint64_t oldIndex) { return oldIndex; }
-
   uint32_t GetMergeWeight(uint64_t fsa_value) { return fsa_value; }
 
-  static value_store_t GetValueStoreType() { return INT_INNER_WEIGHTS_VALUE_STORE; }
-
+  void CloseFeeding() {}
   void Write(std::ostream& stream) const {}
+};
 
-  /**
-   * Close the value store, so no more updates;
-   */
-  void CloseFeeding() {}
+/**
+ * An IntValue store which uses the values as inner weights.
+ */
+class IntInnerWeightsValueStore final : public IntInnerWeightsValueStoreBase {
+ public:
+  explicit IntInnerWeightsValueStore(const keyvi::util::parameters_t& parameters = keyvi::util::parameters_t()) {}
+  IntInnerWeightsValueStore(const IntInnerWeightsValueStore& that) = delete;
+  IntInnerWeightsValueStore& operator=(IntInnerWeightsValueStore const&) = delete;
+};
 
- private:
-  template <typename, typename>
-  friend class ::keyvi::dictionary::DictionaryMerger;
+class IntInnerWeightsValueStoreMerge final : public IntInnerWeightsValueStoreBase {
+ public:
+  explicit IntInnerWeightsValueStoreMerge(const keyvi::util::parameters_t& parameters = keyvi::util::parameters_t()) {}
 
-  uint64_t GetValue(const char* p, uint64_t v, bool* no_minimization) { return v; }
+  IntInnerWeightsValueStoreMerge& operator=(IntInnerWeightsValueStoreMerge const&) = delete;
+  IntInnerWeightsValueStoreMerge(const IntInnerWeightsValueStoreMerge& that) = delete;
+  uint64_t AddValueMerge(const char* p, uint64_t v, bool* no_minimization) { return v; }
+};
+
+class IntInnerWeightsValueStoreAppendMerge final : public IntInnerWeightsValueStoreBase {
+ public:
+  explicit IntInnerWeightsValueStoreAppendMerge(
```

Comment:
Is this constructor needed for append merger ?

Reply:
yes, because the generator constructs ValueStore if it isn't passed into it, similar to the AddValue problem. We should look into this in 1 change.